tests/utils/test_autolog_hparams.py

```diff
@@ -0,0 +1,175 @@
+
+    parsed_dict = extract_hparams(locals_dict)
+    assert parsed_dict == expected_parsed_dict
+
+
+def test_extract_hparams_trainer():
+    train_dl = DataLoader(RandomClassificationDataset(), batch_size=16)
+    model = SimpleModel()
+    optimizer = Adam(model.parameters(), eps=1e-3)
+    trainer = Trainer(
+        model=model,
+        train_dataloader=train_dl,
+        optimizers=optimizer,
+        auto_log_hparams=True,
+        progress_bar=False,
+        log_to_console=False,
+        run_name='test',
+        seed=3,
+        algorithms=[EMA()],
+        loggers=[MLFlowLogger()],
+        callbacks=[SpeedMonitor()],
+    )
+
+    expected_hparams = {
+        'model': 'SimpleModel',
+
+        # Train Dataloader
+        'train_dataloader': 'DataLoader',
+        'train_dataloader_label': 'train',
+        'train_subset_num_batches': -1,
+
+        # Stopping Condition
+        'max_duration': None,
+
+        # Algorithms
+        'algorithms': ['EMA'],
+
+        # Engine Pass Registration
+        'algorithm_passes': None,
+
+        # Optimizers and Scheduling
+        'optimizers': 'Adam',
+        'schedulers': None,
+        'scale_schedule_ratio': 1.0,
+        'step_schedulers_every_batch': None,
+
+        # Evaluators
+        'eval_dataloader': None,
+        'eval_interval': 1,
+        'eval_subset_num_batches': -1,
+
+        # Callbacks and Logging
+        'callbacks': ['SpeedMonitor'],
+        'loggers': ['MLFlowLogger'],
+        'run_name': 'test',
+        'progress_bar': False,
+        'log_to_console': False,
+        'console_stream': 'stderr',
+        'console_log_interval': '1ep',
+        'log_traces': False,
+        'auto_log_hparams': True,
+
+        # Load Checkpoint
+        'load_path': None,
+        'load_object_store': None,
+        'load_weights_only': False,
+        'load_strict_model_weights': False,
+        'load_progress_bar': True,
+        'load_ignore_keys': None,
+        'load_exclude_algorithms': None,
+
+        # Save Checkpoint
+        'save_folder': None,
+        'save_filename': 'ep{epoch}-ba{batch}-rank{rank}.pt',
+        'save_latest_filename': 'latest-rank{rank}.pt',
+        'save_overwrite': False,
+        'save_interval': '1ep',
+        'save_weights_only': False,
+        'save_num_checkpoints_to_keep': -1,
+
+        # Graceful Resumption
+        'autoresume': False,
+
+        # DeepSpeed
+        'deepspeed_config': None,
+        'fsdp_config': None,
+
+        # System/Numerics
+        'device': 'DeviceCPU',
+        'precision': 'Precision',
+        'grad_accum': 1,
+        'device_train_microbatch_size': None,
+
+        # Reproducibility
+        'seed': 3,
+        'deterministic_mode': False,
+
+        # Distributed Training
+        'dist_timeout': 1800.0,
+        'ddp_sync_strategy': None,
+
+        # Profiling
+        'profiler': None,
+
+        # Python logging
+        'python_log_level': None,
+        'auto_microbatching': False,
```

Comment:
It is a bit weird that these local variables from the trainer init function show up...but I guess its fine?

Reply:
I think it's fine. we could move the auto-logging to the top of __init__ before those variables are created? wdyt?

Reply:
It is probably fine either way. If it stays here than any modifications that the trainer init makes to inputs will be captured by the logging. That is probably a good thing.

Reply:
ok sg